ext/semian_cb_data/semian_cb_data.c

```diff
@@ -0,0 +1,900 @@
+static void semian_cb_data_delete_memory_inner (semian_cb_data *ptr);
+static VALUE semian_cb_data_delete_memory (VALUE self);
+static VALUE semian_cb_data_get_successes(VALUE self);
+static VALUE semian_cb_data_set_successes(VALUE self, VALUE num);
+static VALUE semian_cb_data_semid(VALUE self);
+static VALUE semian_cb_data_shmid(VALUE self);
+static VALUE semian_cb_data_array_at_index(VALUE self, VALUE idx);
+static VALUE semian_cb_data_array_set_index(VALUE self, VALUE idx, VALUE val);
+static VALUE semian_cb_data_array_length(VALUE self);
+static VALUE semian_cb_data_set_push_back(VALUE self, VALUE num);
+static VALUE semian_cb_data_set_pop_back(VALUE self);
+static VALUE semian_cb_data_set_push_front(VALUE self, VALUE num);
+static VALUE semian_cb_data_set_pop_front(VALUE self);
+
+// needed for TypedData_Make_Struct && TypedData_Get_Struct
+static const rb_data_type_t
+semian_cb_data_type = {
+  "semian_cb_data",
+  {
+    semian_cb_data_mark,
+    semian_cb_data_free,
+    semian_cb_data_memsize
+  },
+  NULL, NULL, RUBY_TYPED_FREE_IMMEDIATELY
+};
+
+/*
+ * Generate key
+ */
+static key_t
+generate_key(const char *name)
+{
+  union {
+    unsigned char str[SHA_DIGEST_LENGTH];
+    key_t key;
+  } digest;
+  SHA1((const unsigned char *) name, strlen(name), digest.str);
+  /* TODO: compile-time assertion that sizeof(key_t) > SHA_DIGEST_LENGTH */
+  return digest.key;
+}
+
+/*
+ * Log errors
+ */
+static void
+raise_semian_syscall_error(const char *syscall, int error_num)
+{
+  rb_raise(eSyscall, "%s failed, errno %d (%s)", syscall, error_num, strerror(error_num));
+}
+
+/*
+ * Functions that handle type and memory
+*/
+static void
+semian_cb_data_mark(void *ptr)
+{
+  /* noop */
+}
+
+static void
+semian_cb_data_free(void *ptr)
+{
+  semian_cb_data *data = (semian_cb_data *) ptr;
+
+
+  // Under normal circumstances, memory use should be in the order of bytes,
+  //   and shouldn't increase if the same key/id is used
+  //   so there is no need to call this unless certain all other semian processes are stopped
+  //   (also raises concurrency errors: "object allocation during garbage collection phase")
+
+  //semian_cb_data_delete_memory_inner (data);
+
+  xfree(data);
+}
+
+static size_t
+semian_cb_data_memsize(const void *ptr)
+{
+  return sizeof(semian_cb_data);
+}
+
+static VALUE
+semian_cb_data_alloc(VALUE klass)
+{
+  VALUE obj;
+  semian_cb_data *ptr;
+
+  obj = TypedData_Make_Struct(klass, semian_cb_data, &semian_cb_data_type, ptr);
+  return obj;
+}
+
+
+
+
+
+
+/*
+ * Init function exposed as ._initialize() that is delegated by .initialize()
+ */
+static VALUE
+semian_cb_data_init(VALUE self, VALUE id, VALUE size, VALUE permissions)
+{
+  semian_cb_data *ptr;
+  TypedData_Get_Struct(self, semian_cb_data, &semian_cb_data_type, ptr);
+
+  if (TYPE(id) != T_SYMBOL && TYPE(id) != T_STRING)
+    rb_raise(rb_eTypeError, "id must be a symbol or string");
+  if (TYPE(size) != T_FIXNUM /*|| TYPE(size) != T_BIGNUM*/)
```

Comment:
Isn't `size` the the [`error count threshold`](https://github.com/Shopify/semian/blob/circuit-breaker-per-host/lib/semian/circuit_breaker_shared_data.rb#L4)? Should never be larger than 31 bits.